block_writer/main.go

```diff
@@ -51,6 +47,146 @@ var outputInterval = flag.Duration("output-interval", 1*time.Second, "Interval o
 
+// SqlLogic implements basic blockwriter logic for a specific SQL dialect.
+type SqlLogic interface {
+	InsertBlock(int64, uint64, string, []byte) error
+	Setup(string) error
+}
+
+// CockroachLogic implements blockwriter logic for Cockroach SQL.
+type CockroachLogic struct {
+	db *sql.DB
+}
+
+func (cl CockroachLogic) InsertBlock(blockID int64, blockNum uint64, writerID string, data []byte) error {
+	const insertBlockStmt = `INSERT INTO datablocks.blocks (block_id, block_num, writer_id, raw_bytes) VALUES ($1, $2, $3, $4)`
+	_, err := cl.db.Exec(insertBlockStmt, blockID, blockNum, writerID, data)
+	return err
+}
+
+func (cl *CockroachLogic) Setup(connString string) error {
+	parsedURL, err := url.Parse(connString)
+	if err != nil {
+		return err
+	}
+
+	// Remove any database specified in the connection URL.
+	q := parsedURL.Query()
+	q.Del("dbname")
+	parsedURL.RawQuery = q.Encode()
+	parsedURL.RawPath = ""
+
+	// In order to create initial database, open connection to server with no database.
+	cl.db, err = sql.Open("postgres", parsedURL.String())
+	if err != nil {
+		return err
+	}
+	if _, err = cl.db.Exec("CREATE DATABASE IF NOT EXISTS datablocks"); err != nil {
+		return err
+	}
+	if err := cl.db.Close(); err != nil {
+		return err
+	}
+
+	// Open connection to the datablocks database.
+	q = parsedURL.Query()
+	q.Add("dbname", "datablocks")
+	parsedURL.RawQuery = q.Encode()
+	cl.db, err = sql.Open("postgres", parsedURL.String())
+	if err != nil {
+		return err
+	}
+
+	// Create the 'blocks' table, dropping it if it already existed.
+	if _, err := cl.db.Exec(`DROP TABLE IF EXISTS blocks`); err != nil {
+		return err
+	}
+	if _, err := cl.db.Exec(`
+		CREATE TABLE IF NOT EXISTS blocks (
+		  block_id BIGINT NOT NULL,
+		  writer_id TEXT NOT NULL,
+		  block_num BIGINT NOT NULL,
+		  raw_bytes BYTES NOT NULL,
+		  PRIMARY KEY (block_id, writer_id, block_num)
+		)`); err != nil {
+		return err
+	}
+
+	cl.db.SetMaxOpenConns(*concurrency + 1)
+	return nil
+}
+
+// CockroachLogic implements blockwriter logic for Postgres SQL.
+type PostgresLogic struct {
+	db *sql.DB
+}
+
+func (pl PostgresLogic) InsertBlock(blockID int64, blockNum uint64, writerID string, data []byte) error {
+	const insertBlockStmt = `INSERT INTO blocks (block_id, block_num, writer_id, raw_bytes) VALUES ($1, $2, $3, $4)`
+	_, err := pl.db.Exec(insertBlockStmt, blockID, blockNum, writerID, data)
+	return err
+}
+
+func (pl *PostgresLogic) Setup(connString string) error {
+	parsedURL, err := url.Parse(connString)
+	if err != nil {
+		return err
+	}
+
+	// Remove any dbname specified in the provided connection url.
+	q := parsedURL.Query()
+	q.Set("dbname", "postgres")
+	parsedURL.RawQuery = q.Encode()
+	parsedURL.RawPath = ""
+
+	// Open connection to the 'postgres' database.
+	pl.db, err = sql.Open("postgres", parsedURL.String())
+	if err != nil {
+		return err
+	}
+	// There is no conditional database creation in postgres; we must first
```

Comment:
There's no conditional creation, but there is a conditional drop, so you could do `DROP DATABASE IF EXISTS datablocks` and then an unconditional create.